### PR TITLE
Add time evolution driver with jax-based ODE solver

### LINKED ISSUE #289: Adding inexact time evolution for neural quantum states
Big fan of this project, and the paper(s) that have inspired it! This is less an issue than a question. If I understand correctly, the original Carleo & Troyer (2017) Science paper goes into some detail about doing time evolution on a neural network state. I've had some trouble implementing this on my own, however, to replicate the results of that paper. I'd love to see that functionality added to NetKet, but I know that may not be in the works. If not, perhaps I could talk to one of the authors (who seem to be the drivers of this project) about how the time evolution was implemented for their paper. I am interested in building on the method for a research project.

### LINKED ISSUE #435: Time-evolution: new design
Now that everything is pythonically [~ugly~](https://www.youtube.com/watch?v=BCQpitbo-SA) nice, writing a variational time integrator that plugs into scipy is quite easy. I would like to take the occasion to unify the interface among time evolution and standard drivers, so this is an issue to discuss this.

Requirements:
- The interface should be consistent with the vmc optimization drivers as much as possible
- The interface should support open and hamiltonian evolution, otherwise [people](https://github.com/netket/netket/graphs/contributors) will break my beloved lindblad without me noticing 
- Should reuse other package's time-integrators. I [am lazy](https://en.wikipedia.org/wiki/Sleep) and don't like to for the millionth time a RK integrator
- Should be somewhat easy to plug another package integrator in, if needed.

Wishes:
 - I like the idea of it being easy to first define a problem (machine, initial state, time interval) and then solve it repeatedly with different solvers/timesteps/options. For anyone, this is essentially the [DiffEq/SciML interface](https://docs.sciml.ai/dev/tutorials/ode_example/), where you first define a problem and then you solve it. I don't know if this is feasible but that would be nice.
```julia
f(u,p,t) = 1.01*u
u0 = 1/2
tspan = (0.0,1.0)
prob = ODEProblem(f,u0,tspan)
sol = solve(prob, rk45(), reltol=1e-8, abstol=1e-8, dt=0.001)
``` 

Conceptually, this means:
 - Define as usual the machine, the sampler
 - You *MUST* define the SR object at some point, because it is always required
 - Define the (potentially time dependent) hamiltonian/lindblad you wish to integrate.
 - Define the integrator, and all it's related options (in order to be generic, those arguments should be forwarded...)
- Define the interval in time we wish to integrate upon.

possible implementation:
in order to be somewhat portable, we need to be able to output a lambda function that takes a flattened vector of parameters and computes the gradient, which is equivalent to defining the function that defines an ode:
```python
def create_odefun(op, sampler, *args, **kwargs):
    machine = sampler._machine
    if not isinstance(machine, AbstractDensityMatrix):
        driver = Vmc(op, sampler, *args, optimizer=None, **kwargs)
    else:
        driver = SteadyState(op, sampler, *args, optimizer=None, **kwargs)

    def _fun(t, w):
        machine.parameters = unflatten(w)
        dp = driver._time_evo_step()
        return flatten(dp)

   return _fun, driver
```

It would be a worthwhile discussion whever this `time_evo_step` method should live in `vmc.py` and `steadystate.py` or it should be a separate class, or just a standalone method that does dispatch.

Having this function I can already plug into scipy, diffeqpy or any other integrator module easily.
Everything else is fancy logging and parameter handling. My main preoccupation for this is that we want to compute observables only ever once in a while (otherwise it really gets slow fast). And unfortunately scion does not allow callback hooks like that.

--

Since we like to be fancy, we would like to have an higher level `TimeEvolution` driver.
An initial proposal might be something along the lines of the following. 
```python
H = ...
ma = nk.machine.rbmspin(hilbert=hi, alpha=1)
ma.parameters = my_physical_state
sa = nk.sampler.MetropolisLocal(machine=ma, n_chains=8)
sr = nk.optimizer.SR(diag_shift=0.001, use_iterative=True, sparse_tol=1e-6, sparse_maxiter=1000)

tevo = TimeEvolution(H, sa, sr=sr, n_samples=2000) # incase of a density matrix also, sampler_obs=sa_obs, n_samples_obs=500)
tevo.solver("rk45", t_0=(1.0, 2.0), dt=0.01, adaptive=False)

tevo.run(1.0, out="test", obs=obs, step_size=0.02) # goes from 1.0 to 2.0, saving logs every 0.02
 48%|██▎      | 0.4800/1.0 [05:28<18:19, 723.68s/time, Energy=-25.3 ± 0.01 [var=4.6e+01, R_hat=0.9993]] 
pars_end = copy(ma.parameters)


# now if I want to test out a different tilmestep I should do...
ma.parameters = my_physical_state
tevo.solver("rk45", t_0=1.0, dt=0.005, adaptive=False)
tevo.run(1.0, out="test", obs=obs, step_size=0.02) # goes from 1.0 to 2.0, saving logs every 0.02
...

# that's better! now I continue my integration to t=3.0
tevo.run(1.0, out="test", obs=obs, step_size=0.02) # goes from 2.0 to 3.0, saving logs every 0.02
...
```

I already implemented the above, and it's nice that it uses the same `run` interface of Abstractdriver with only minimal changes. Ill admit this is not the best interface ever. But it allows to have a nice driver interface while also being able to call `iterate` manually for only a few iterations if one so whishes.

One thing I don't like is that if you randomly interrupt this, you will have random parameters for a random Time stored in the machine you have referenced outside... but this is a core design of netket, so I'm not sure we can do much about it.

Alternatively we might more go for something like
```python
H = ...
ma = nk.machine.rbmspin(hilbert=hi, alpha=1)
ma.parameters = my_physical_state
sa = nk.sampler.MetropolisLocal(machine=ma, n_chains=8)
sr = nk.optimizer.SR(diag_shift=0.001, use_iterative=True, sparse_tol=1e-6, sparse_maxiter=1000)

tevo_prob = TimeEvolutionProblem(H, sa, time_interval, sr=sr, n_samples=2000)
final_state, logged_obs = solve(tevo_prob, "rk45", dt=0.01, adaptive=False, obs=obs)

# want to try out other dt
final_state_dt2, logged_obs_dt2 = solve(tevo_prob, "rk45", dt=0.005, adaptive=False, obs=obs)

# that's better! now let's continue the time evolution...
tevo_prob_2 = TimeEvolutionProblem(H, sa, time_interval_2, sr=sr, n_samples=2000, state= final_state_dt2)
# state= will simply assign to the machine..
final_state_3, logged_obs_3 = solve(tevo_prob_2, "rk45", dt=0.005, adaptive=False, obs=obs)
```

I like this because it splits conceptually the time integrator (the solver) from the definition of the problem itself. However the first one is simpler to use in general if you don't play around much with solving the same thing with different solvers, but only integrate once.

In my experience when working with time evolution one does many tests at first, but then settles for one thing....
So i dunno.
What do you think?

Miscellaneous stuff:
- SciPy does not support fixed time-step integration! 
   - kinda. But I've been playing around recently with it, and if you set `abstol, reltol= ∞` and `max_dt`, effectively you are fixing the dt to that value. It works well.
- I don't like your jokes
   - Meh. me neither.

### PR DESCRIPTION
**Pull request by femtobit**
This PR adds an implementation of t-VMC based on the synthesis of @PhilipVinc's and my ideas for t-VMC propagation using `jax`. This is added to the `netket.experimental` submodule for the moment, to give us more freedom to continue improving the API (in particular of the included ODE solver routines).

This PR contains two main parts:

1.  A jax-based implementation of Runge-Kutta integration schemes (both adaptive and with fixed step size) in `netket.experimental.dynamics.runge_kutta`. These are not tied to t-VMC or even variational quantum states, but can in principle be applied to any explicit ODE defined as a Python function `dx/dt = f(x, t)`. The code is written with the goal of making the whole solver jitable by `jax` (although this is not enabled in this PR, since our VMC code is not yet jitable.)
2. The `netket.experimental.TDVP` driver, which supports t-VMC propagation of pure NQS (that is, `MCState`) based on either the real-time or imaginary-time Schrödinger equation. Lindbladian real-time dynamics for NDMs using `MCMixedState` are also supported.

Basic usage of the driver looks like this:
```python3
integrator = nkx.dynamics.RK23(dt=0.01, adaptive=True, rtol=1e-3, atol=1e-3)
te = nkx.TDVP(
    hamiltonian,
    variational_state=vstate,
    integrator=integrator,
)
te.run(
    T=1.0,
    out="example_dynamics",
    tstops=np.linspace(0.0, 1.0, 101, endpoint=True),
)
```
See the included example (`Examples/Dynamics/ising1d.py`) for a complete demo.
There are also tests under `test/dynamics` that show some of the parts of this PR in action.

Closes #289, closes #435.

### CODE REVIEW

**Comment by gcarleo** [comment]
thanks a lot @femtobit and @PhilipVinc ! before going into the serious stuff, let me make the first superficial comments on names: 
I would call ``` integrator_config``` just ```integrator```, and I am not sure ```tstops``` is a good name, is this the checkpoint times? 

Other than that, just to understand the flexbility of the implementation a bit better: do we have (already) a mechanism to potentially support time dependent hamiltonians? 

Concerning this: 
```python 
 qgt=nk.optimizer.qgt.QGTJacobianDense(holomorphic=True)
```

I was wondering if we can completely remove any reference to the optimizer namespace (which is a bit confusing in this context of dynamics), and just call the QGT of the variational state instead? (at least in the example maybe)

**Comment by femtobit** [comment]
>  I would call integrator_config just integrator, and I am not sure tstops is a good name, is this the checkpoint times?

I agree, just `integrator` should be fine. After all, we do say just `qgt` too, not `qgt_config` or so.

If `tstops` are passed, they are the times where the integrator computes observables, does logging, calls the passed callback etc. or, in `TimeDependentVMC.iter` the points where the loop body is called:
```python3
>>> ts = []
>>> for t in te.iter(T=1.0, tstops=[0.1, 0.2, 0.314, 0.79]): ts.add(t)
>>> print(ts)
[0.1, 0.2, 0.314, 0.79]
```
The solver makes sure to step to those points by temporarily using a smaller time step when necessary. This behavior is similar to what the [`tstops` option(https://diffeq.sciml.ai/stable/basics/common_solver_opts/#Output-Control) in DifferentialEquations.jl] does, hence the name.

> Other than that, just to understand the flexbility of the implementation a bit better: do we have (already) a mechanism to potentially support time dependent hamiltonians?

Time-dependent Hamiltonians are already possible. For example:
```python3
te = nkx.TimeDependentVMC(hamiltonian, lambda t: nk.operator.Ising(..., h=np.sin(t))
```

> I was wondering if we can completely remove any reference to the optimizer namespace (which is a bit confusing in this context of dynamics), and just call the QGT of the variational state instead? (at least in the example maybe)

Since the `QGT...` objects are not an actual QGT but more like a template, that is only turned into a concrete QGT when pased to `VariationalState.quantum_geoemtric_tensor`.

I agree that the `optimizer` namespace is not ideal. We could think about moving the QGT interface to `netket.vqs.qgt`, since the concept of the QGT is closely tied to the variational parametrization and, as we've noticed recently in #953, the implementation of the QGTs also is closely tied to the implementation of `VariationalState`. That'd actually a pretty nice solution, in my opinion.

By the way: The only reason we have to specify the QGT is the example at all is because the default QGT settings print a warning that performance is not optimal for holomorphic NQS unless `holomorphic=True` is passed (which would not be a good look in the example).

File: netket/experimental/dynamics/runge_kutta.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+def scaled_error(y, y_err, atol, rtol, *, last_norm_y=None, norm_fn):
+    norm_y = norm_fn(y)
+    scale = (atol + jnp.maximum(norm_y, last_norm_y) * rtol) / nkjax.tree_size(y_err)
+    return norm_fn(y_err) / scale, norm_y
+
+
+LimitsType = Tuple[Optional[float], Optional[float]]
+
+
+def propose_time_step(dt, scaled_error, limits: LimitsType):
```
Comment (PhilipVinc) [inline netket/experimental/dynamics/runge_kutta.py:194]: can you add a comment on where you took this from?

**Review by PhilipVinc — COMMENTED** [review]
i also agree that `integrator_config` should be renamed to `integrator`

I also dislike `TimeDependentVMC` as a name, it's very verbose.

Note: we should find a way to allow for changing the tilmestep controller.

**Comment by femtobit** [comment]
> I also dislike TimeDependentVMC as a name, it's very verbose.

It is. I am open to suggestions. I wanted to avoid something like `tVMC` since it starts with a lower-case letter and `TimeEvolution` is very generic (maybe if our `VMC` driver was still called `GroundState`...).

> Note: we should find a way to allow for changing the tilmestep controller.

We'd just need to extract a generic interface and add a `step_controller=...` option to the integrator. That is doable.

File: netket/experimental/driver/tvmc.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+        else:
+            self._generator = operator
+
+        self.propagation_type = propagation_type
+        if propagation_type == "real":
+            self._loss_grad_factor = -1.0j
+        elif propagation_type == "imag":
+            self._loss_grad_factor = -1.0
+        else:
+            raise ValueError("propagation_type must be one of 'real', 'imag'")
```
Comment (PhilipVinc) [inline netket/experimental/driver/tvmc.py:127]:
```python
        self.propagation_type = propagation_type
        if isinstance(variational_state, VariationalMixedState):
            # assuming Lindblad Dynamics
            # TODO: support density matrix imaginary time evolution
            if propagation_type == "real":
                self._loss_grad_factor = 1.0j
            else:
                raise ValueError("only real-time Lindblad evolution is supported for "
                                 "mixed states")
        else:
            if propagation_type == "real":
                self._loss_grad_factor = -1.0j
            elif propagation_type == "imag":
                self._loss_grad_factor = -1.0
            else:
                raise ValueError("propagation_type must be one of 'real', 'imag'")

```

File: netket/experimental/driver/tvmc.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
+        self.qgt = qgt
+        self.linear_solver = linear_solver
+        self.linear_solver_restart = linear_solver_restart
+
+        self._w = self.state.parameters
+        self._dw = None  # type: PyTree
+        self._last_qgt = None
+
+        self.integrator = integrator
```
Comment (PhilipVinc) [inline netket/experimental/driver/tvmc.py:137]:
```suggestion
```

File: netket/experimental/driver/tvmc.py (hunk 2 above)
Comment (femtobit) [reply netket/experimental/driver/tvmc.py:127]: This alone won't work because there is no method `dwdt(_: MCMixedState, ...)` to dispatch to.

Comment (PhilipVinc) [reply netket/experimental/driver/tvmc.py:127]: yes it will work.
MCMixedState is a MCState

Comment (PhilipVinc) [reply netket/experimental/driver/tvmc.py:127]: that's why i check it before

Comment (femtobit) [reply netket/experimental/driver/tvmc.py:127]: >  MCMixedState is a MCState

Ah, I didn't realize that.

File: netket/experimental/driver/tvmc.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        operator: AbstractOperator,
+        variational_state: VariationalState,
+        integrator_config: RKIntegratorConfig,
+        *,
+        t0: float = 0.0,
+        propagation_type="real",
+        qgt: LinearOperator = None,
+        linear_solver=None,
+        linear_solver_restart: bool = False,
+        error_norm: str = "euclidean",
```
[the remark this answers is not in the document]
Comment (femtobit) [reply netket/experimental/driver/tvmc.py]: This is supported now.

File: netket/experimental/driver/tdvp.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+from netket.utils import mpi
+from netket.utils.dispatch import dispatch
+from netket.utils.types import PyTree
+from netket.vqs import VariationalState, VariationalMixedState, MCState
+
+from netket.experimental.dynamics import RKIntegratorConfig
+
+
+@dispatch
+def dwdt(state, driver, t, w, *, stage=None):
```
Comment (PhilipVinc) [inline netket/experimental/driver/tdvp.py:41]: this in theory is part of the interface (probably not, but still...) so maybe we should find a better name?

File: netket/experimental/driver/tdvp.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    # pylint: disable=unused-argument
+    raise NotImplementedError(f"dwdt not implemented for {type(state)}")
+
+
+@dispatch
+def dwdt_mcstate(state: MCState, driver, t, w, *, stage: int = None):
```
Comment (PhilipVinc) [inline netket/experimental/driver/tdvp.py:47]:
```suggestion
def dwdt(state: MCState, driver, t, w, *, stage: int = None):
```

File: netket/experimental/driver/tdvp.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+                ("state         ", self.state),
+            ]
+        ]
+        return "\n{}".format(" " * 3 * (depth + 1)).join([str(self)] + lines)
+
+    def _odefun(self, t, w, **kwargs):
+        """
+        Internal method which dispatches to the actual ODE system function.
+        """
+        return dwdt_mcstate(self.state, self, t, w, **kwargs)
```
Comment (PhilipVinc) [inline netket/experimental/driver/tdvp.py:387]:
```suggestion
        return dwdt(self.state, self, t, w, **kwargs)
```

or any better name we come up with, really.

File: netket/experimental/driver/tdvp.py (hunk 5 above)
Comment (femtobit) [reply netket/experimental/driver/tdvp.py:41]: Let's keep `dwdt` for the time being. We can change it later, if we come up with a better name.

**Review by PhilipVinc — APPROVED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,8 @@
...
 * {ref}`nk.operator.LocalOperator` now accepts sparse matrices as input operators [#919](https://github.com/netket/netket/pull/919)
 * A new variational state that performs exact summation over the whole Hilbert space has been added. It can be constructed with {ref}`nk.vqs.ExactState` and supports the same Jax neural networks as {ref}`nk.vqs.MCState`. [#953](https://github.com/netket/netket/pull/953)
+* A new time-evolution driver that uses the Time-dependent-variational-principle has been added to the experimental submodule as {ref}`nk.experimental.TDVP`. It works with time-independent and time-dependent hamiltonians and liouvillians. [#1012](https://github.com/netket/netket/pull/1012) 
+* A set of Runge-Kutta, jax-compatible integrators has been added to the experimental module to use together with the new TDVP driver. [#1012](https://github.com/netket/netket/pull/1012)
```
Comment (femtobit) [inline CHANGELOG.md:18]:
```suggestion
* [Experimental] A new time-evolution driver  {ref}`nk.experimental.TDVP` using the time-dependent variational principle (TDVP) has been added. It works with time-independent and time-dependent Hamiltonians and Liouvillians. [#1012](https://github.com/netket/netket/pull/1012) 
* [Experimental] A set of JAX-compatible Runge-Kutta ODE integrators has been added for use together with the new TDVP driver. [#1012](https://github.com/netket/netket/pull/1012) 
```

File: docs/docs/api-experimental.rst [hunk 9]
```diff
@@ -62,3 +62,36 @@ Variational State Interface
...
+This is a collection of ODE integrators that can be used with the TDVP 
+driver above.
+
+.. currentmodule:: netket
+
+.. autosummary::
+  :toctree: _generated/experimental/dynamics
+  :nosignatures:
+
+
```
Comment (femtobit) [inline docs/docs/api-experimental.rst:90]:
```suggestion
```

File: docs/docs/api-experimental.rst [hunk 10]
```diff
@@ -62,3 +62,36 @@ Variational State Interface
...
+  nkx.dynamics.Euler
+  nkx.dynamics.Heun
+  nkx.dynamics.Midpoint
+  nkx.dynamics.RK12
+  nkx.dynamics.RK23
+  nkx.dynamics.RK4
+  nkx.dynamics.RK45
```
Comment (femtobit) [inline docs/docs/api-experimental.rst:97]:
```suggestion
  nkx.dynamics.RK45
  
```

File: netket/experimental/driver/tdvp.py [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+    ):
+        r"""
+        Initializes the time evolution driver.
+
+        Args:
+            operator: The generator of the dynamics (Hamiltonian for pure states,
+                Lindbladian for density operators).
+            variational_state: The variational state.
+            integrator: Configuration of the algorithm used for solving the ODE.
+            t0: Initial time.
```
Comment (femtobit) [inline netket/experimental/driver/tdvp.py:97]:
```suggestion
            t0: Initial time at the start of the time evolution.
```

File: netket/experimental/driver/tdvp.py [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+        # Essentially means
+        return None
+
+    @property
+    def step_value(self):
+        return self.t
+
+    @property
+    def dt(self):
+        return self._integrator.dt
```
Comment (femtobit) [inline netket/experimental/driver/tdvp.py:359]:
```suggestion
    def dt(self):
        """Current time step."""
        return self._integrator.dt
```

File: netket/experimental/driver/tdvp.py [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+
+    @property
+    def t(self):
+        return self._integrator.t
```
Comment (femtobit) [inline netket/experimental/driver/tdvp.py:363]:
```suggestion
    @property
    def t(self):
        """Current time."""
        return self._integrator.t
```

File: docs/docs/api-experimental.rst [hunk 14]
```diff
@@ -62,3 +62,36 @@ Variational State Interface
...
+  :toctree: _generated/experimental/dynamics
+  :nosignatures:
+
+  nkx.dynamics.Euler
+  nkx.dynamics.Heun
+  nkx.dynamics.Midpoint
+  nkx.dynamics.RK12
+  nkx.dynamics.RK23
+  nkx.dynamics.RK4
+  nkx.dynamics.RK45
```
Comment (PhilipVinc) [inline docs/docs/api-experimental.rst:96]:
```suggestion
  netket.experimental.dynamics.Euler
  netket.experimental.dynamics.Heun
  netket.experimental.dynamics.Midpoint
  netket.experimental.dynamics.RK12
  netket.experimental.dynamics.RK23
  netket.experimental.dynamics.RK4
  netket.experimental.dynamics.RK45
```

**Review by PhilipVinc — APPROVED** [review]

File: test/utils/test_utils.py [hunk 15]
```diff
@@ -61,3 +62,17 @@ def test_HashableArray(numpy):
...
+    ksum1 = KahanSum(0.0)
+    ksum2 = KahanSum(0.0)
+    vals = 0.01 * np.ones(500)
+    for val in vals:
+        sum += val
+        ksum1 += val
+        ksum2 = ksum2 + val
+
+    assert ksum1.value == pytest.approx(5.0, rel=1e-15, abs=1e-15)
+    assert ksum2.value == pytest.approx(5.0, rel=1e-15, abs=1e-15)
```
Comment (femtobit) [inline test/utils/test_utils.py:78]:
```suggestion
def test_Kahan_sum():
    ksum1 = KahanSum(0.0)
    ksum2 = KahanSum(0.0)
    vals = 0.01 * np.ones(500)
    for val in vals:
        ksum1 += val
        ksum2 = ksum2 + val

    assert ksum1.value == pytest.approx(5.0, rel=1e-15, abs=1e-15)
    assert ksum2.value == pytest.approx(5.0, rel=1e-15, abs=1e-15)
```

File: netket/utils/summation.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        delta = other - self.compensator
+        new_value = self.value + delta
+        new_compensator = (new_value - self.value) - delta
+        return KahanSum(new_value, new_compensator)
+
+    def __radd__(self, other: Scalar):
+        delta = other - self.compensator
+        new_value = self.value + delta
+        self.compensator = (new_value - self.value) - delta
+        self.value = new_value
```
Comment (femtobit) [inline netket/utils/summation.py:41]:
```suggestion
    def __iadd__(self, other: Scalar):
        delta = other - self.compensator
        new_value = self.value + delta
        self.compensator = (new_value - self.value) - delta
        self.value = new_value
```

File: Examples/Dynamics/ising1d.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+# Quenched hamiltonian: this has a different transverse field than `ha`
+ha1 = nk.operator.Ising(hilbert=hi, graph=g, h=0.5)
+te = nkx.TDVP(
+    ha1,
+    variational_state=vs,
+    integrator=integrator,
+    t0=0.0,
+    qgt=nk.optimizer.qgt.QGTJacobianDense(holomorphic=True),
+    error_norm="qgt",
```
Comment (gcarleo) [inline Examples/Dynamics/ising1d.py:66]: maybe I am mistaken, but I cannot find docs for error_norm

File: netket/experimental/driver/tdvp.py [hunk 18]
```diff
@@ -0,0 +1,405 @@
...
+            variational_state: The variational state.
+            integrator: Configuration of the algorithm used for solving the ODE.
+            t0: Initial time at the start of the time evolution.
+            propagation_type: Determines the equation of motion: "real"  for the
+                real-time Schödinger equation (SE), "imag" for the imaginary-time SE.
+            qgt: The QGT specification.
+            linear_solver: The solver for solving the linear system determining the time evolution.
+            linear_solver_restart: If False (default), the last solution of the linear system
+                is used as initial value in subsequent steps.
+            error_norm: Norm function used to calculate the error with adaptive integrators.
```
Comment (gcarleo) [inline netket/experimental/driver/tdvp.py:104]: sorry my bad... I found it

File: Examples/Dynamics/ising1d.py (hunk 17 above)
Comment (PhilipVinc) [reply Examples/Dynamics/ising1d.py:66]: It's here https://www.netket.org/preview/dh/tevo-jax/docs/_generated/experimental/dynamics/netket.experimental.driver.TDVP.html#netket.experimental.driver.TDVP

**Comment by femtobit** [comment]
>  can you change the test so it fails if it does not support inplace addition?

It did support it. Python just uses `__add__` if no custom `__iadd__` is there (which would probably also be fine). I just noticed that the custom method was not called due to the coverage.

**Pull request merged**